lighting/raymarch/volume.glsl

```diff
@@ -1,103 +1,137 @@
+            if (mat.sdf <= 0.0) {
+                return vec3(0.0, 0.0, 0.0);
             }
-            col += vec4(LIGHT_COLOR * 80. * tmp * T * Tl, 1.0);
+        #endif
+        Medium resL = RAYMARCH_VOLUME_MAP_FNC(positionL);
+        float densityL = -resL.sdf;
+        vec3 extinctionL = resL.absorption + resL.scattering;
+        transmittanceL *= exp(-densityL * extinctionL * stepSizeL);
+
+        float offset = random(position)*(stepSizeL*RAYMARCH_VOLUME_DITHER);
+        tL += stepSizeL + offset;
+    }
+
+    return transmittanceL;
+}
+
+vec3 raymarchVolume(vec3 rayOrigin, vec3 rayDirection, vec2 st, float minDist, vec3 background) {
+    vec3 scatteredLuminance = vec3(0.0, 0.0, 0.0);
+    vec3 transmittance = vec3(1.0, 1.0, 1.0);
+    float stepSize = RAYMARCH_MAX_DIST/float(RAYMARCH_VOLUME_SAMPLES);
+
+    float t = RAYMARCH_MIN_DIST;
+
+    for(int i = 0; i < RAYMARCH_VOLUME_SAMPLES; i++) {        
+        vec3 position = rayOrigin + rayDirection * t;
+        Medium res = RAYMARCH_VOLUME_MAP_FNC(position);
+        float density = -res.sdf;
+        vec3 extinction = res.absorption + res.scattering;
+
+        if (t < minDist && density > 0.0) {
+
+            #if defined(LIGHT_DIRECTION) || defined(LIGHT_POSITION)
+                #if defined(LIGHT_DIRECTION) // directional light
+                    float stepSizeL = RAYMARCH_MAX_DIST/float(RAYMARCH_VOLUME_SAMPLES_LIGHT);
+                    vec3 rayDirectionL = normalize(LIGHT_DIRECTION);
+                    const float attenuationL = 1.0;
+                #else // point light
+                    float distL = distance(LIGHT_POSITION, position);
+                    float stepSizeL = distL/float(RAYMARCH_VOLUME_SAMPLES_LIGHT);
+                    vec3 rayDirectionL = normalize(LIGHT_POSITION - position);
+                    float attenuationL = attenuation(distL);
+                #endif
+                vec3 shadow = shadowTransmittance(position, rayDirectionL, stepSizeL);
+                vec3 L = LIGHT_COLOR * LIGHT_INTENSITY;
+            #else // no lighting
+                const float attenuationL = 1.0;
+                const vec3 shadow = 1.0;
+                const vec3 L = vec3(1.0, 1.0, 1.0);
+            #endif
+
+            #if defined RAYMARCH_ENERGY_CONSERVING
+                // energy-conserving scattering integration
+                vec3 S = L * attenuationL * shadow * density * res.scattering;
+                vec3 sampleExtinction = max(vec3(EPSILON, EPSILON, EPSILON), density * extinction);
+                vec3 Sint = (S - S * exp(-sampleExtinction * stepSize)) / sampleExtinction;
+                scatteredLuminance += transmittance * Sint;
+                transmittance *= exp(-sampleExtinction * stepSize);
+            #else
+                // usual scattering integration. Not energy-conserving.
+                scatteredLuminance += attenuationL * shadow * transmittance * density * res.scattering * stepSize * L;
```

Comment:
```
scatteredLuminance += attenuationL * shadow * transmittance * density * scattering * stepSize * L;
```

Reply:
@patriciogonzalezvivo same as above. `res.scattering` looks correct to me. `scaterring` is not defined.

Reply:
Ah ok now I see.